packages/yew/src/virtual_dom/key.rs

```diff
@@ -1,74 +1,189 @@
-}
+#[derive(Debug, Clone, Copy, Ord, PartialOrd, Eq, PartialEq, Hash)]
+#[repr(transparent)]
+pub struct Key(self::imp::Repr);
 
-impl Display for Key {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        self.key.fmt(f)
+// Default implementation for byte sequences.
+impl From<&[u8]> for Key {
+    #[inline]
+    fn from(key: &[u8]) -> Self {
+        Self(self::imp::from_hash(BYTES_SEED, key))
     }
 }
 
-impl Deref for Key {
-    type Target = str;
-
-    fn deref(&self) -> &str {
-        self.key.as_ref()
+impl From<Rc<[u8]>> for Key {
+    #[inline]
+    fn from(key: Rc<[u8]>) -> Self {
+        Key::from(key.as_ref())
     }
 }
 
 impl From<Rc<str>> for Key {
+    #[inline]
     fn from(key: Rc<str>) -> Self {
-        Self { key }
+        Self::from(key.as_ref().as_bytes())
     }
 }
 
-impl From<&'_ str> for Key {
-    fn from(key: &'_ str) -> Self {
-        let key: Rc<str> = Rc::from(key);
-        Self::from(key)
+impl From<Arc<[u8]>> for Key {
+    #[inline]
+    fn from(key: Arc<[u8]>) -> Self {
+        Key::from(key.as_ref())
+    }
+}
+
+impl From<Arc<str>> for Key {
+    #[inline]
+    fn from(key: Arc<str>) -> Self {
+        Self::from(key.as_ref().as_bytes())
+    }
+}
+
+impl From<&str> for Key {
+    #[inline]
+    fn from(key: &str) -> Self {
+        Self::from(key.as_bytes())
+    }
+}
+
+impl From<String> for Key {
+    #[inline]
+    fn from(key: String) -> Self {
+        Self::from(key.as_bytes())
+    }
+}
+
+impl<const N: usize> From<[u8; N]> for Key {
+    #[inline]
+    fn from(key: [u8; N]) -> Self {
+        Self::from(key.as_ref())
+    }
+}
+
+impl From<Vec<u8>> for Key {
+    #[inline]
+    fn from(key: Vec<u8>) -> Self {
+        Self::from(key.as_slice())
+    }
+}
+
+impl From<char> for Key {
+    #[inline]
+    fn from(key: char) -> Key {
+        Self::from(key as u32)
     }
 }
 
 impl ImplicitClone for Key {}
 
-macro_rules! key_impl_from_to_string {
+impl From<u128> for Key {
+    #[inline]
+    fn from(key: u128) -> Self {
+        Self(self::imp::from_hash(UNSIGNED_SEED, key))
+    }
+}
+
+impl From<i128> for Key {
+    #[inline]
+    fn from(key: i128) -> Self {
+        Self(self::imp::from_hash(SIGNED_SEED, key))
+    }
+}
+
+macro_rules! unsigned {
     ($type:ty) => {
         impl From<$type> for Key {
+            #[inline]
             fn from(key: $type) -> Self {
-                Self::from(key.to_string().as_str())
+                Self::from(key as u128)
             }
         }
     };
 }
 
-key_impl_from_to_string!(String);
-key_impl_from_to_string!(char);
-key_impl_from_to_string!(u8);
-key_impl_from_to_string!(u16);
-key_impl_from_to_string!(u32);
-key_impl_from_to_string!(u64);
-key_impl_from_to_string!(u128);
-key_impl_from_to_string!(usize);
-key_impl_from_to_string!(i8);
-key_impl_from_to_string!(i16);
-key_impl_from_to_string!(i32);
-key_impl_from_to_string!(i64);
-key_impl_from_to_string!(i128);
-key_impl_from_to_string!(isize);
+macro_rules! signed {
```

Comment:
Does this mean that `Key::from(0_u8) == Key::from(0_u16)` but `Key::from(0_i8) != Key::from(0_u8)`?

Reply:
Yeah, this can be remedied with zig zag encoding [as mentioned under **Solving the shortcomings** here](https://github.com/yewstack/yew/pull/2616#issue-1205585356).

Reply:
Personally, I feel that passing a key of `0_u16` into an element / component where its last key is `0_u8` to be treated as the same element / component may not be an intentional behaviour for most cases.

I would prefer to treat them as different values if a correct implementation could potentially negatively impact code size.

We can emit a warning if the type of key changes between renders in debug mode.